File: cs/example/Samples/Group.cs
/*
 * File: Group.cs
 * Project: Test
 * Created Date: 23/05/2021
 * Author: Shun Suzuki
 * -----
 * Last Modified: 08/03/2023
 * Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
 * -----
 * Copyright (c) 2022 Shun Suzuki. All rights reserved.
 * 
 */


using AUTD3Sharp;
using AUTD3Sharp.Gain;
using AUTD3Sharp.Gain.Holo;
using AUTD3Sharp.Modulation;
using AUTD3Sharp.Utils;

namespace Samples;

internal static class GroupTest
{
    public static void Test(Controller autd)
    {
        var config = new SilencerConfig();
        autd.Send(config, TimeSpan.FromMilliseconds(20));

        var g1 = new Focus(autd.Geometry.CenterOf(0) + new Vector3d(0, 0, 150));
        var g2 = new GSPAT();
        g2.Add(autd.Geometry.CenterOf(1) + new Vector3d(30.0, 0.0, 0.0), 1.0);
        g2.Add(autd.Geometry.CenterOf(1) - new Vector3d(30.0, 0.0, 0.0), 1.0);

        var gain = new Grouped();
        gain.Add(0, g1);
        gain.Add(1, g2);
        var mod = new Sine(150); // AM sin 150 Hz
        autd.Send(mod, gain, TimeSpan.FromMilliseconds(20));
    }
}

File: python/example/samples/group.py
"""
File: group.py
Project: samples
Created Date: 15/09/2023
Author: Shun Suzuki
-----
Last Modified: 15/09/2023
Modified By: Shun Suzuki (suzuki@hapis.k.u-tokyo.ac.jp)
-----
Copyright (c) 2023 Shun Suzuki. All rights reserved.

"""


import numpy as np

from pyautd3 import Controller, Device, Silencer
from pyautd3.gain import Focus, Group, Null
from pyautd3.modulation import Sine, Static


async def group_by_device(autd: Controller) -> None:
    config = Silencer()
    await autd.send_async(config)

    def grouping(dev: Device) -> str | None:
        match dev.idx:
            case 0:
                return "null"
            case 1:
                return "focus"
            case _:
                return None

    await (
        autd.group(grouping)
        .set_data("null", Static(), Null())
        .set_data(
            "focus",
            Sine(150),
            Focus(autd.geometry.center + np.array([0.0, 0.0, 150.0])),
        )
        .send_async()
    )


async def group_by_transducer(autd: Controller) -> None:
    config = Silencer()
    await autd.send_async(config)

    cx = autd.geometry.center[0]
    g1 = Focus(autd.geometry.center + np.array([0, 0, 150]))
    g2 = Null()

    g = Group(lambda _, tr: "focus" if tr.position[0] < cx else "null").set_gain("focus", g1).set_gain("null", g2)

    m = Sine(150)

    await autd.send_async((m, g))
